### PR TITLE
signalflow fakebackend support for timestamp control and END_OF_CHANNEL

### PR DESCRIPTION
**Pull request by jherbage**
Timestamps are introduced via the execute request counting up in resolutionMs increments between the startMs time and the stopMs time. When the stopMs is reached the END_OF_CHANNEL is reached.
This makes the fakebackend moure usuable in unit tests since we can have deterministic timestamps and the client can disconnect when the computation all expected datapoints.
Unit tests added to demonstrate this works with multiple clients connecting.

### CODE REVIEW

File: signalflow/fake_backend.go [hunk 1]
```diff
@@ -177,6 +177,26 @@ func (f *FakeBackend) handleMessage(ctx context.Context, message map[string]inte
...
+
+		if stopMs == 0 {
+			stopMs = uint64(time.Now().UnixNano() / (1000 * 1000))
+		}
+		messageStartMs, _ := message["start"].(float64)
+		if messageStartMs != 0.0 {
+			startMs = uint64(messageStartMs)
+		}
+
+		if stopMs == 0 {
```
Comment (keitwb) [inline signalflow/fake_backend.go:196]: How could this ever be 0 since if it is 0 above on line 188 it gets set to the current time in ms?

Comment (jherbage) [reply signalflow/fake_backend.go]: its a typo - fixed so startMs is now set to now in line with how the client default works

File: signalflow/fake_backend.go [hunk 2]
```diff
@@ -177,6 +177,26 @@ func (f *FakeBackend) handleMessage(ctx context.Context, message map[string]inte
...
+			startMs = uint64(time.Now().Add(-1*time.Minute).UnixNano() / (1000 * 1000))
```
[the remark this answers is not in the document]
Comment (jherbage) [reply signalflow/fake_backend.go]: defaults for stopMs and startMs now aligned to flow client
stopMs defaults to 0 which means the computation doesnt end
startMs defaults to now

File: signalflow/client.go [hunk 3]
```diff
@@ -300,6 +300,7 @@ func (c *Client) Stop(req *StopRequest) error {
 	// connection is still active (i.e. we don't need to call ensureInitialized
 	// here).  If the websocket connection does drop, all jobs started by that
 	// connection get stopped automatically.
+	fmt.Println("stop")
```
Comment (keitwb) [inline signalflow/client.go:303]: Remove this.

File: signalflow/conn.go [hunk 4]
```diff
@@ -119,6 +119,7 @@ func (c *wsConn) Run() {
 			conn = nil
 			time.Sleep(ReconnectDelay)
 		case msg := <-c.outgoingTextMsgs:
+			fmt.Println(string(msg.bytes))
```
Comment (keitwb) [inline signalflow/conn.go:122]: Remove this.

**Comment by jherbage** [comment]
Closing this as messed up the commit history on my fork
Opened pull request 145 to replace it

**Pull request closed without merging**